### PR TITLE
Initial work towards cublasXt wrappers

### PR DESCRIPTION
**Pull request by kshyatt**
Something is going wrong after the tests run, all tests after the `cublasXt` ones fail due to the library not being initialized and I'm having a hell of a time figuring out why.

### CODE REVIEW

**Comment by kshyatt** [comment]
bors try

**Comment by kshyatt** [comment]
OK, so reordering the tests "worked" (and I fixed a broken one in CUSOLVER). Something about cublasXt and CUSOLVER makes them not want to play nicely.

**Comment by kshyatt** [comment]
bors try

**Comment by maleadt** [comment]
> OK, so reordering the tests "worked" (and I fixed a broken one in CUSOLVER). Something about cublasXt and CUSOLVER makes them not want to play nicely.

Well that is disconcerting. Could you reproduce locally? You mentioned something about creating a new context on Slack, could you elaborate?

**Comment by kshyatt** [comment]
Yes, I can reproduce locally. If the CUBLAS tests run last everyone is happy but they seem to corrupt something for all later tests, and only when the `Xt` tests are run (thus, only when `cublasXtCreate` is called...).

**Comment by kshyatt** [comment]
OK, updated again ... I think something is wrong with the finalizer for `xt` which would explain why everything worked when the BLAS tests run last. All the test failures seem to be in CUBLAS calls post the `xt` tests BEFORE I tried to reset the context here, it's almost all https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverSPstatus INTERNAL_ERROR aka `async_timeout`. So I think there must be something funny happening to the primary context.

**Comment by maleadt** [comment]
Did some debugging, looks like the cublasXt initialization mutates the active context:

```
CUDAdrv.CuCurrentContext() = CUDAdrv.CuContext(Ptr{Nothing} @0x00000000024bd030, true, true)
(CuArrays.CUBLAS).cublasXtDeviceSelect(handle, 1, Cint[0]) = 0x00000000
CUDAdrv.CuCurrentContext() = CUDAdrv.CuContext(Ptr{Nothing} @0x0000000001f86f30, false, true)
```

This of course messes with everything.

**Comment by maleadt** [comment]
This is annoying:

```julia
using CuArrays
using CUDAdrv

println("Initial context:      ", repr(CUDAdrv.CuCurrentContext()))

CUDAdrv.Mem.alloc(1)

println("Context after init:   ", CUDAdrv.CuCurrentContext())

handle = CuArrays.CUBLAS.cublasXtCreate()

println("Context after create: ", CUDAdrv.CuCurrentContext())

CuArrays.CUBLAS.cublasXtDeviceSelect(handle, 1, Cint[0])

println("Context after select: ", CUDAdrv.CuCurrentContext())
```

```
Initial context:      nothing
Context after init:   CuContext(Ptr{Nothing} @0x0000000002daa2c0, true, true)
Context after create: CuContext(Ptr{Nothing} @0x0000000002daa2c0, true, true)
Context after select: CuContext(Ptr{Nothing} @0x0000000002323a00, false, true)
```

```c++
#include <stdio.h>

#include <cublasXt.h>
#include <cuda_runtime.h>
#include <cuda.h>

int main(int argc, char **argv) {
  CUcontext ctx = NULL;
  cuCtxGetCurrent(&ctx);
  printf("Initial context:      %p\n", ctx);

  void *dummy;
  cudaMalloc(&dummy, 1);

  cuCtxGetCurrent(&ctx);
  printf("Context after init:   %p\n", ctx);

  cublasXtHandle_t handle;
  if (cublasXtCreate(&handle) != CUBLAS_STATUS_SUCCESS) {
    fprintf(stderr, "!!!! CUBLASXT initialization error\n");
    return EXIT_FAILURE;
  }

  cuCtxGetCurrent(&ctx);
  printf("Context after create: %p\n", ctx);

  int devices[] = {0};
  if (cublasXtDeviceSelect(handle, 1, devices) != CUBLAS_STATUS_SUCCESS) {
    fprintf(stderr, "!!!! CUBLASXT device selection error\n");
    return EXIT_FAILURE;
  }

  cuCtxGetCurrent(&ctx);
  printf("Context after select: %p\n", ctx);

  if (cublasXtDestroy(handle) != CUBLAS_STATUS_SUCCESS) {
    fprintf(stderr, "!!!! shutdown error (A)\n");
    return EXIT_FAILURE;
  }
}
```

```
Initial context:      (nil)
Context after init:   0x555f3e254c20
Context after create: 0x555f3e254c20
Context after select: 0x555f3e254c20
```

**Comment by maleadt** [comment]
OK, seems like we need to use the primary context mechanism. Looking into this.

**Comment by kshyatt** [comment]
What the heck. Am I reading that right that it only mutates stuff in the Julia version, and only after device selection? Weeeeeeird.

I'm "glad" my suspicion about context funny business turned out to be right 😛

**Comment by maleadt** [comment]
> What the heck. Am I reading that right that it only mutates stuff in the Julia version, and only after device selection? Weeeeeeird.

Yeah, it seems like cublasXt uses the active _primary_ context, and allocates a new one if it doesn't exist. We were using regular contexts, so initializing cublasXt replaced our context with a primary one.
https://github.com/JuliaGPU/CUDAnative.jl/pull/360 should fix this.

**Comment by maleadt** [comment]
This passes tests locally, back to you @kshyatt :slightly_smiling_face: 
bors try

**Comment by kshyatt** [comment]
wooo! DO NOT merge this, I have more tests to write

**Comment by kshyatt** [comment]
bors try

**Comment by kshyatt** [comment]
OK, I think this is mergeable if others think it looks good.

File: test/blas.jl [hunk 1]
```diff
@@ -1380,6 +1379,298 @@ end
     @test C ≈ h_C
 end
 
-end
+@testset "xt" begin
+    @testset "trmm! with element type $elty" for elty in [Float32, Float64, ComplexF32] # test broken for ComplexF64
```
Comment (vchuravy) [inline test/blas.jl]: What part is broken here? Is it a tolerance issue?

Comment (kshyatt) [reply test/blas.jl]: Some of the off diagonal elements are wrong (and not by a little), I think.

File: src/blas/CUBLAS.jl [hunk 2]
```diff
@@ -35,6 +37,24 @@ function handle()
...
 end
 
+function xt_handle()
+    if _xt_handle[] == C_NULL
+        @assert isassigned(active_context) # some other call should have initialized CUDA
+        _xt_handle[] = get!(_xt_handles, active_context[]) do
+            context = active_context[]
+            dev = CUDAdrv.device(context)
+            handle = cublasXtCreate()
+            cublasXtDeviceSelect(handle, 1, [dev.handle])
```
Comment (maleadt) [inline src/blas/CUBLAS.jl]: why not default to using all available devices? isn't that the whole point of cublasXt?
it seems like you can't switch modes afterwards (or at least not between calls), so it'd need to happen here.

>  Currently the device configuration is static and cannot be changed between Math function calls. In that regard, this function should be called only once after cublasXtCreate. To be able to run multiple configurations, multiple cublasXt API contexts should be created.

File: test/blas.jl (hunk 1 above)
Comment (kshyatt) [reply test/blas.jl]: I tried this locally and it works now??? Weird.

File: src/blas/CUBLAS.jl [hunk 3]
```diff
@@ -35,6 +37,24 @@ function handle()
...
+            #cublasXtSetBlockDim(handle, 64)
```
Comment (maleadt) [inline src/blas/CUBLAS.jl]: I guess this can go? This call can happen later.

**Comment by kshyatt** [comment]
bors try

File: test/sparse_solver.jl [hunk 4]
```diff
@@ -81,6 +81,7 @@ k = 1
 
     @testset "csreigvsi" begin
         A     = sparse(rand(elty,n,n))
+        A     = A + A'
```
Comment (maleadt) [inline test/sparse_solver.jl]: didn't tests pass before this change?

File: test/runtests.jl [hunk 5]
```diff
@@ -22,6 +21,7 @@ testf(f, xs...; kwargs...) = GPUArrays.TestSuite.compare(f, CuArray, xs...; kwar
 
 allowscalar(false)
 
+using Test
```
Comment (maleadt) [inline test/runtests.jl]: maybe get rid of this change (unless it has a reason?)

File: test/sparse_solver.jl (hunk 4 above)
Comment (kshyatt) [reply test/sparse_solver.jl]: Not on my machine

File: src/blas/CUBLAS.jl (hunk 2 above)
Comment (kshyatt) [reply src/blas/CUBLAS.jl]: I set this for testing purposes and forgot about it! I'll try adding in more devices too :)

**Comment by kshyatt** [comment]
bors try

**Comment by kshyatt** [comment]
Yaaaaay! Thanks, @maleadt!

**Pull request merged**